src/AutoRest.CSharp/Common/Input/InputTypes/Serialization/TypeSpecInputModelTypeConverter.cs

```diff
@@ -135,58 +135,74 @@ private static InputModelType CreateInputModelTypeInstance(string? id, string? n
-                 */
-                if (property != null && isMultipartType)
+                var rawProperty = reader.ReadWithConverter<InputModelProperty>(options);
+                if (rawProperty == null)
+                {
+                    throw new JsonException($"Property of model '{modelName}' cannot be null.");
+                }
+
+                if (rawProperty.IsFlattened)
                 {
-                    static InputType ConvertPropertyType(InputType propertyType)
+                    if (isMultipartType)
                     {
-                        return propertyType switch
-                        {
-                            InputPrimitiveType { Kind: InputPrimitiveTypeKind.Bytes } => InputPrimitiveType.Stream,
-                            InputListType listType => new InputListType(listType.Name, listType.CrossLanguageDefinitionId, ConvertPropertyType(listType.ValueType))
-                            {
-                                Decorators = listType.Decorators
-                            },
-                            InputDictionaryType dictionaryType => new InputDictionaryType(dictionaryType.Name, dictionaryType.KeyType, ConvertPropertyType(dictionaryType.ValueType))
-                            {
-                                Decorators = dictionaryType.Decorators
-                            },
-                            InputNullableType nullableType => new InputNullableType(ConvertPropertyType(nullableType.Type))
-                            {
-                                Decorators = nullableType.Decorators
-                            },
-                            _ => propertyType
-                        };
+                        throw new JsonException($"Flattened property '{rawProperty.Name}' cannot be used in a multipart type '{modelName}'.");
                     }
 
-                    property = new InputModelProperty(property.Name,
-                                                    property.SerializedName,
-                                                    property.Description,
-                                                    ConvertPropertyType(property.Type),
-                                                    property.ConstantValue,
-                                                    property.IsRequired,
-                                                    property.IsReadOnly,
-                                                    property.IsDiscriminator,
-                                                    property.FlattenedNames);
+                    var flattenedProperties = FlattenProperty(rawProperty);
+
+                    foreach (var property in flattenedProperties)
+                    {
+                        properties.Add(property);
+                    }
+                }
+                else
+                {
+                    properties.Add(isMultipartType ? ConvertMultipartProperty(rawProperty) : rawProperty);
```

Comment:
here we have an issue about the optionality of the properties.
In modeler v4, if the `properties` is optional, all the flattened properties will be optional, but here we are not doing anything with their optionality.
This issue exists before in the old flatten properties code in the emitter, but we did not notice until today.

Reply:
it does not have impacts right now (or maybe but we did not notice), because in almost all typespec input MPG specs, the flatten is not enabled. But we need to fix this.

Reply:
We discussed offline. If the flattened property is optional, we prefer to make all the properties in that property optional.

Reply:
It's easy to fix the property definition, but it's not easy to fix our serialization logic. I think it would be better to resolve it in another issue.

Reply:
filed https://github.com/Azure/autorest.csharp/issues/5101 for follow-up discussion